### DISCUSSION TITLE
Persist states are inconsistent when used in components

### DISCUSSION THREAD

**Discussion by Esd4g** [pmndrs/zustand#1552] [Q&A] [answered]
## Summary
I have a store that looks like this:
```
export const useSlurmStateStore = create(
    persist(
        set => ({
            values: ["mixed", "drain", "drained", "down"],
            setValues: (newValues) => set({ values: newValues })
        }),
        {
            name: 'slurm-state',
            storage: createJSONStorage(() => localStorage)
        }
    )
)
```
It is used in a MUI-based component like this:
```
export default function CamieFilterSlurmState() {

  const [values, setValues] = useSlurmStateStore((state) => [state.values, state.setValues])

  const handleChange = (event, newFormats) => {
    setValues(newFormats);
  };

  return (
    <ToggleButtonGroup
      color="primary"
      value={values}
      onChange={handleChange}
    >
      <ToggleButton value="idle">Idle</ToggleButton>
      <ToggleButton value="mixed">Mixed</ToggleButton>
      <ToggleButton value="drain">Drain</ToggleButton>
      <ToggleButton value="drained">Drained</ToggleButton>
      <ToggleButton value="down">Down</ToggleButton>
    </ToggleButtonGroup>
  )
}
```
The value is also accessed on a page like this:
```
const valuesSlurmState = useSlurmStateStore((state) => state.values)
```
When I reload the page, the values on the page are loaded from localStorage, while the values in the component are loaded from the default values defined in the store. Everything else works as expected.

Is this zustand's fault? Am I doing something wrong?

**Comment by dai-shi** [comment]
> When I reload the page, the values on the page are loaded from localStorage, while the values in the component are loaded from the default values defined in the store.

I'm not sure if I follow. What are those two kinds of values?

**Reply by Esd4g** [reply]
Sorry for not providing much information.
values is an array of strings, which determines whether ToggleButtons inside the ToggleButtonGroup are selected or not. If the array includes a string, the button with the corresponding value is set. So an empty array means no buttons are selected and ["idle", "mixed", "drain", "drained", "down"] means all buttons are selected.
values is used on a page to only show elements whose SlurmState is included in values (= via buttons I can choose which elements not to show based on their SlurmState).
When first accessing the page, values is set to what it is in the store, but after that it is supposed to change through user input and persist, so that it stays the same when reloading/loading a different page with the same filter functionality.

**Reply by dai-shi** [reply]
Hmm, can you create a very small reproduction using https://codesandbox.io/s/react-new ?

**Reply by Esd4g** [reply]
As it turns out, this problem doesn't occur in React.js:
https://codesandbox.io/s/nifty-hoover-t1jfy7
But it does occur in Next.js, which my original project is running with:
https://stackblitz.com/edit/nextjs-vggxab

I guess it's not zustand's responsibility then?

**Reply by dai-shi** [reply]
perhaps.

just to confirm, you don't see any issues without `persist` middleware, right?

**Reply by Esd4g** [reply]
Without persist I'd expect "values" to default to ["mixed", "drain", "drained", "down"] every time I reload the page, which it does.

Is it useful to create a bug report on the next.js repo for this?

**Reply by dai-shi** [reply]
Okay, unfortunately your example is still too big and I don't get if it's an issue in zustand or not.
If it were an issue in zustand with nextjs, it'd be probably related to this: #1537 

If it were an issue in nextjs, you would probably need to make a reproduction without zustand to file a bug report in nextjs repo.

**Reply by Esd4g** [reply]
The examples should now contain the bare minimum, though I doubt that's small enough. Shouldn't matter, I found multiple small adjustments to my code that would produce a hydration error, so it's most likely related to #1537.

I am certainly not skilled enough to reproduce this phenomenon without zustand.

**Comment by Esd4g** [comment, marked as answer]
Solution to my problem:
https://stackoverflow.com/questions/72649255/nextjs-zustand-persist-state
